numpy/lib/tests/test_polynomial.py

```diff
@@ -136,27 +136,39 @@ def test_polyfit(self):
         weights = np.arange(8, 1, -1)**2/7.0
 
         # Check exception when too few points for variance estimate. Note that
-        # the Bayesian estimate requires the number of data points to exceed
-        # degree + 3.
+        # the estimate requires the number of data points to exceed
+        # degree + 1
         assert_raises(ValueError, np.polyfit,
-                      [0, 1, 3], [0, 1, 3], deg=0, cov=True)
+                      [1], [1], deg=0, cov=True)
+
+        # Check exception when option absolute_weights is True, but  no weights
+        # are given
+        assert_raises(ValueError, np.polyfit,
+                      x, y, deg=1, absolute_weights=True)
 
         # check 1D case
```

Comment:
Could we add a "check 0D case"? I ask since for that case we know exactly what the answer should be. (True also for higher-D, but less obvious): just pass in normally distributed data and check that `sigma_mean = sqrt(cov) = sigma/sqrt(n)`. E.g.,
```
np.random.seed(123)
y = np.random.normal(size=(4, 10000))
mean, cov = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True)
# Should get sigma_mean = sigma/sqrt(N) = 1./sqrt(4) = 0.5.
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.5, atol=0.01)
``` 
(Obviously, the last test would fail on current master, which goes to show that the -2 is wrong at least for the simplest case.)

Reply:
Just realized the above is missing a test of the unscaled version:
```
mean, cov_noscaling = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True, cov_scale=False)
# without scaling, for 0D: <cov> = (N-dof)/N = 3/4.
assert_allclose(cov.mean(), 0.75, atol=0.01)
```

Reply:
Sorry, but I'm not sure if I understand the tests correctly. My calculations go as follows:
1) absolute_weights = False:
var(a) = var(y) / n_dof
(biased) variance of drawing 4 Gaussian random numbers with width 1: var_4 = (n-1)/n*var = 3/4
therefore: var(a)= var_4 / (n-1) = var / n 
or: sigma(a) = sqrt(sigma/n) = 1/sqrt(4) = 1/2

2) absolute_weights = True:
ar(a) = 1/S, with S=sum_{i=1}^n (1/var_i) 
with var_1 = var_2 = ... var_n = var: S = n / var
or: var(a) = var / n
then: sigma(a) = sqrt(var/ n)  
with var=1: sigma_1(a) = sqrt(1/4) = 1/2
with var=3/4: sigma_3/4(a) = sqrt(3/4/4) = sqrt(3)/4 = 0.43301270...

For the relative weights (1) we seem to agree while for the absolute weights (2), there is a difference:

```python
import numpy as np

np.random.seed(123)
y = np.random.normal(size=(4, 10000))
x = np.zeros(y.shape[0])

print("*** relative weights ***")
mean, cov = np.polyfit(x, y, deg=0, cov=True, absolute_weights=False)
print("expected: 1/sqrt(4)=0.5")
print("np.sqrt(cov.mean())={}".format(np.sqrt(cov.mean())))
print("")
print("")
print("*** absolute weights***")
print("a) var_i=1")
var = np.ones(y.shape[0])
mean, cov = np.polyfit(
    x, y, deg=0, cov=True, w=1 / np.sqrt(var), absolute_weights=True)
print("expected: 1/sqrt(4)=0.5")
print("np.sqrt(cov.mean())={}".format(np.sqrt(cov.mean())))
print("")

print("b) var_i=3/4")
var = np.full(y.shape[0], 3 / 4)
mean, cov = np.polyfit(
    x, y, deg=0, cov=True, w=1 / np.sqrt(var), absolute_weights=True)
print("expected: sqrt(3)/4={}".format(np.sqrt(3) / 4))
print("np.sqrt(cov.mean())={}".format(np.sqrt(cov.mean())))
```
gives
```
*** relative weights ***
expected: 1/sqrt(4)=0.5
np.sqrt(cov.mean())=0.5008913441736218


*** absolute weights***
a) var_i=1
expected: 1/sqrt(4)=0.5
np.sqrt(cov.mean())=0.5

b) var_i=3/4
expected: sqrt(3)/4=0.4330127018922193
np.sqrt(cov.mean())=0.43301270189221924
```

Reply:
Yes, I'm an idiot! For the unscaled case, of course, the `sqrt(cov.mean())` again should be 0.5, since the weights are 1 and by construction the reduced chi2 is 1 too.

So, I think the tests might be better as:
```
np.random.seed(123)
y = np.random.normal(size=(4, 10000))
mean, cov = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True)
# Should get sigma_mean = sigma/sqrt(N) = 1./sqrt(4) = 0.5.
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.5, atol=0.01)
# Without scaling, since reduced chi2 is 1, the result should be the same.
mean, cov = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True, absolute_weights=True)
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.5, atol=0.01)
# If we estimate our errors wrong, no change with scaling:
w = np.full(y.shape[0], 1./0.5**2)
mean, cov = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True)
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.5, atol=0.01)
# But if we do not scale, our estimate for the error in the mean will differ.
mean, cov = np.polyfit(np.zeros(y.shape[0], y, deg=0, cov=True, absolute_weights=True)
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.25, atol=0.01)
```

Reply:
I guess, the 3rd test should look like this?

```Python
# If we estimate our errors wrong, no change with scaling:
w = np.full(y.shape[0], 1./0.5**2)
mean, cov = np.polyfit(np.zeros(y.shape[0]), y, w, deg=0, cov=True)
assert_allclosel(mean.std(),  0.5, atol=0.01)
assert_allclose(np.sqrt(cov.mean()), 0.5, atol=0.01)
```
(the weights `w` were passed to np.polyfit)

Reply:
@wummo - yes, that was the idea, sorry for writing tests on the fly, without actually checking them!